lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb

```diff
@@ -0,0 +1,198 @@
+
+          emulators_list = running_emulators.map(&:serial)
+          # Get the intersection of the set of running emulators with the ones we want to shut down
+          emulators_list &= serials unless serials.nil?
+          emulators_list.each do |e|
+            Actions.sh(@tools.adb, '-s', e, 'emu', 'kill') { |_| } # ignore error if no emulator with specified serial is running
+
+            # NOTE: Alternative way of shutting down emulator would be to call the following command instead, which shuts down the emulator more gracefully:
+            # `adb -s #{e} shell reboot -p` # In case you're wondering, `-p` is for "power-off"
+            # But this alternate command:
+            #  - Requires that `-no-snapshot` was used on boot (to avoid being prompted to save current state on shutdown)
+            #  - Disconnects the emulator from `adb` (and thus disappear from `adb devices -l`) for a short amount of time,
+            #    before reconnecting to it but in an `offline` state, until `emulator` finally completely quits and it disappears
+            #    again (for good) from `adb devices --list`.
+            # This means that so if we used alternative, we couldn't really retry_loop until emulator disappears from `running_emulators` to detect
+            # that the shutdown was really complete, as we might as well accidentally detect the intermediate disconnect instead.
+          end
+
+          # Wait until all emulators are killed
+          retry_loop(time_between_retries: SHUTDOWN_WAIT, timeout: SHUTDOWN_TIMEOUT, description: 'waiting for devices to shutdown') do
+            (emulators_list & running_emulators.map(&:serial)).empty?
+          end
+
+          UI.success('All emulators are now shut down.')
+        end
+
+        # Find the system-images package for the provided API, with Google APIs, and matching the current platform/architecture this lane is called from.
+        #
+        # @param [Integer] api The Android API level to use for this AVD
+        # @return [String] The `system-images;android-<N>;google_apis;<platform>` package specifier for `sdkmanager` to use in its install command
+        #
+        # @note Results from this method are memoized, to avoid repeating calls to `sdkmanager` when querying for the same api level multiple times.
+        #
+        def system_image_package(api:)
+          @system_image_packages ||= {}
+          @system_image_packages[api] ||= begin
+            platform = `uname -m`.chomp
+            all_packages = `#{@tools.sdkmanager} --sdk_root=#{@tools.android_sdk_root} --list`
+            package = all_packages.match(/^ *(system-images;android-#{api};google_apis;#{platform}(-[^ ]*)?)/)&.captures&.first
+            UI.user_error!("Could not find system-image for API `#{api}` and your platform `#{platform}` in `sdkmanager --list`. Maybe Google removed it for download and it's time to update to a newer API?") if package.nil?
```

Comment:
Mostly because last time I god a `PANIC: Broken AVD system image` errors when working on screenshots automation and trying to use the system-images that were defined in our scripts, it took me a very long time to understand what was broken (ref: paqN3M-cM-p2#broken-avd-system-image), and get why it suddenly didn't work on my machine while it worked for other team members (who had the system image installed on their machine from them downloading it a long time ago before it got removed from Google's servers), before I finally understand that that the only way to fix that was to update the configuration.

Given Google apparently removes old system images from their download servers from time to time, this makes this screenshot automation lane able to work on machines that installed the system image from a previous download but fail on machines that try to download it for the first time, and make for that lane to work one day, but break if you try to re-run it later without any change, i.e. it depends on Google download servers changes that makes old successful builds (e.g. from CI) not always reproducible for that reason. That's why I thought this suggestion was important to mention in this case.

Reply:
Fair 👌